doc/engineering/coding_standards.md

```diff
@@ -0,0 +1,236 @@
+
+A test should also be submitted to ensure that the bug is not accidentally reintroduced in the future.
+
+### New Functionality/Features Guidelines
+
+If you are implementing a new feature, also write new test cases.  The primary reason for writing test cases is not to prove that the new feature works correctly, but to make sure that it will be noticed if future changes — perhaps to code that seems unrelated — break the feature.
+
+If you are implementing a new feature, also update the documentation to describe the feature.  If the documentation belongs to a missing document, such as an addition to an undocumented API it is acceptable to only document the addition.  It is nice to create the full missing documentation section ;)
+
+Make sure the patch does not break backward compatibility. In general, we only break backward compatibility in major releases and only for a very good reason and usually after first deprecating the feature one or two releases beforehand.
+
+If you must submit an incompatible change we prefer that the incompatible behaviour require an admin to enable the new functionality via a configuration parameter.
+
+### New Applications Guidelines
+
+TODO
+MR: Thoughts: naming, copyrights section, support contacts that should work, !!!security testing!!!
+
+### Core Updates/Modules Guidelines
+
+TODO
+MR: Thoughts: place where stuff goes (wapi, documents, deps, etc etc and how it should be broken down), comments, tests, support obligations
+
+## Be Ego-less
+
+### Understand and accept that you will make mistakes.
+
+The point is to find them early, before they make it into production. Fortunately, except for the few of us developing rocket guidance software at JPL, mistakes are rarely fatal in our industry, so we can, and should, learn, laugh, and move on.
+
+### You are not your code.
+
+Remember that the entire point of a review is to find problems, and problems will be found. Don't take it personally when one is uncovered.
+
+### No matter how much "karate" you know, someone else will always know more.
+
+Such an individual can teach you some new moves if you ask. Seek and accept input from others, especially when you think it's not needed.
+
+### Don't rewrite code without consultation.
+
+There's a fine line between "fixing code" and "rewriting code." Know the difference, and pursue stylistic changes within the framework of a code review, not as a lone enforcer. Don’t use a sledgehammer on a problem to try and drive in a finishing nail - if code isn’t doing what you want and you don’t know how it’s supposed to work, ask first.
+
+### Treat people who know less than you with respect, deference, and patience.
+
+Nontechnical people who deal with developers on a regular basis almost universally hold the opinion that we are prima donnas at best and crybabies at worst. Don't reinforce this stereotype with anger and impatience.
```

Comment:
I believe this is pulled from [here](https://blog.codinghorror.com/the-ten-commandments-of-egoless-programming/) which we should probably link to as well.